pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/BloomFilterRule.java

```diff
@@ -0,0 +1,142 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class  BloomFilterRule extends AbstractRule {
+  private final Logger LOGGER = LoggerFactory.getLogger(BloomFilterRule.class);
+  private final BloomFilterRuleParams _params;
+  protected final BrokerRequestOptimizer _brokerRequestOptimizer = new BrokerRequestOptimizer();
+
+  public BloomFilterRule(InputManager inputManager, ConfigManager outputManager) {
+    super(inputManager, outputManager);
+    _params = inputManager.getBloomFilterRuleParams();
+  }
+
+  @Override
+  public void run() {
+    int numDims = _inputManager.getNumDims();
+    double[] weights = new double[numDims];
+    AtomicDouble totalWeight = new AtomicDouble(0);
+
+    // For each query, find out the dimensions used in 'EQ'
+    // and accumulate the (weighted) frequencies
+    _inputManager.getQueryWeightMap().forEach((query,weight) -> {
+      totalWeight.addAndGet(weight);
+      FixedLenBitset fixedLenBitset = parseQuery(query);
+      LOGGER.debug("fixedLenBitset {}", fixedLenBitset);
+      for (Integer i : fixedLenBitset.getOffsets()) {
+        weights[i] += weight;
+      }
+    });
+    LOGGER.debug("Weight: {}, Total {}", weights, totalWeight);
+
+    for (int i = 0; i < numDims; i++) {
+      String dimName = _inputManager.intToColName(i);
+      if (((weights[i] / totalWeight.get()) > _params.THRESHOLD_MIN_PERCENT_EQ_BLOOMFILTER)
+          //The partitioned dimension should be frequently > P used
+          && (_inputManager.getCardinality(dimName)
+          < _params.THRESHOLD_MAX_CARDINALITY_BLOOMFILTER)) { //The Cardinality < C (1 million for 1MB size)
+        _outputManager.getIndexConfig().getBloomFilterColumns().add(dimName);
+      }
+    }
+  }
+
+  public FixedLenBitset parseQuery(String queryString) {
+    LOGGER.debug("Parsing query: {}", queryString);
+    if (queryString == null) {
+      return FixedLenBitset.IMMUTABLE_EMPTY_SET;
+    }
+
+    BrokerRequest brokerRequest;
+    AbstractCompiler parser = PinotQueryParserFactory.get(_inputManager.getQueryType());
+    try {
+      brokerRequest = parser.compileToBrokerRequest(queryString);
+    } catch (SqlCompilationException e) {
+      LOGGER.error("Error parsing query: {}, {}", queryString, e.getMessage());
+      return FixedLenBitset.IMMUTABLE_EMPTY_SET;
+    }
+    BrokerRequest optimizedRequest = _brokerRequestOptimizer.optimize(brokerRequest, _inputManager.getTimeCol());
+    QueryContext queryContext = BrokerRequestToQueryContextConverter.convert(optimizedRequest);
+
+    if (queryContext.getFilter() == null) {
+      return FixedLenBitset.IMMUTABLE_EMPTY_SET;
+    }
+
+    LOGGER.trace("Parsing Where Clause: {}", queryContext.getFilter().toString());
+    return parsePredicateList(queryContext.getFilter());
+  }
+
+  /**
+   * The partitioned dimension should used in the “=” （IN, NOT IN, != are not using bloom filter in Pinot for now) filter.
+   * @param filterContext filterContext
```

Comment:
Mark this is as a TODO since it can be easily spotted in the code if someone stumbles upon it in the future. 
TODO: once Pinot starts supporting bloom filter based pruning for IN, !=, NOT IN, we should enhance the algorithm of this rule.

Reply:
done